lib/beautiful/gtk.lua

```diff
@@ -0,0 +1,169 @@
+---------------------------------------------------------------------------
+--- Querying current GTK+ 3 theme via GtkStyleContext.
+--
+-- @author Yauheni Kirylau &lt;yawghen@gmail.com&gt;
+-- @copyright 2016-2017 Yauheni Kirylau
+-- @module beautiful.gtk
+---------------------------------------------------------------------------
+local gears_debug = require("gears.debug")
+local gears_math = require("gears.math")
+local join = require("gears.table").join
+local unpack = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
+
+
+local gtk = {
+  cached_theme_variables = nil
+}
+
+
+local function convert_gtk_channel_to_hex(channel_value)
+    return string.format("%02x", gears_math.round(channel_value * 255))
+end
+
+local function convert_gtk_color_to_hex(gtk_color)
+  return "#" ..
+    convert_gtk_channel_to_hex(gtk_color.red) ..
+    convert_gtk_channel_to_hex(gtk_color.green) ..
+    convert_gtk_channel_to_hex(gtk_color.blue) --..
+  --convert_gtk_channel_to_hex(gtk_color.alpha)
+end
+
+local function lookup_gtk_color_to_hex(_style_context, color_name)
+  local gtk_color = _style_context:lookup_color(color_name)
+  if not gtk_color then
+    return nil
+  end
+  return convert_gtk_color_to_hex(gtk_color)
+end
+
+local function get_gtk_color_property_to_hex(_style_context, property_name)
+    local state = _style_context:get_state()
+    local property = _style_context:get_property(property_name, state)
+    if not property then
+      return nil
+    end
+    local result = convert_gtk_color_to_hex(property.value)
+    property:unset()
+    return result
+end
+
+local function read_gtk_color_properties_from_widget(gtk_widget, properties)
+  local _style_context = gtk_widget:get_style_context()
+  local result = {}
+  for result_key, style_context_property in pairs(properties) do
+    result[result_key] = get_gtk_color_property_to_hex(
+        _style_context, style_context_property
+    )
+  end
+  return result
+end
+
+
+--- Get GTK+3 theme variables from GtkStyleContext
+-- @treturn table subj
+function gtk.get_theme_variables()
+  if gtk.cached_theme_variables then
+    return gtk.cached_theme_variables
+  end
+
+  local result = {}
+  local lgi = require('lgi')
+  local Gtk = lgi.Gtk
+  local window
+  pcall(function()
+    window = Gtk.Window{
+      --on_destroy = Gtk.main_quit,
+    }
+  end)
+  if not window then
+    gears_debug.print_warning("Seems like GTK+3 theme is not set correctly or luajit was built with an incomptible GTK+3 version.")
+    return nil
+  end
```

Comment:
How about replacing the above with (since your answer to this comment last time said you did not understand what I meant):
```lua
local lgi = require("lgi")
local Gtk, window = require("gears.protected_call").call(function()
    local Gtk = lgi.Gtk
    return Gtk, Gtk.Window{}
end)
if not Gtk then
    gears.debug.print_warning("Failed to load GTK and instantiate a GtkWindow.")
    return nil
end
```
(Can/should the `require`s be moved to the beginning of the file?)
(No, I will not ask since when luajit depends against Gtk in some way)

Reply:
>  since when luajit depends against Gtk in some way
i was very sleepy when writing that, i meant `lua-lgi`, not `luajit`

i thought not to move requires to the beginning of file in order not to fire them unless function is called explicitly